[Off-chain] feat: in-memory query cache(s)

bryanchriswhite opened the pull request:
## Summary

Adds the `QueryCache[T any]` and `HistoricalQueryCache[T any]` interfaces, `InMemoryCache[T any]` implementation, configurations, and options functions.

## Issue

<!-- READ & DELETE:
     - Explain the reasoning for the PR in 1-2 sentences. Adding a screenshot is fair game.
     - If applicable: specify the ticket number below if there is a relevant issue; _keep the `-` so the full issue is referenced._
-->

- #543

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,245 @@
+
+	versions := heightMap.(map[int64]heightCacheItem[T])
+	var nearestHeight int64 = -1
+	for h := range versions {
```
Prune and/or ensure a value older than `pruneOlderThan` isn't returned.

Olshansk commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,69 @@
+	}
+}
+
+// WithMaxKeys sets the maximum number of distinct key/value pairs the cache will
```
TOL: MaxEntries sounds pretty bad but that's what this is, no? Just curious what you think.

bryanchriswhite replied:
I considered `MaxEntries` but didn't feel like it was as clear as `MaxKeys`. It would be more idiomatic if we were in a different language (e.g. javascript, python, ruby, C#, etc.). :sweat_smile:

Olshansk commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,69 @@
+	}
+}
+
+// WithTTL sets the time-to-live for cached items. Items older than the TTL
```
` Items older than the TTL MAY not be evicted but SHOULD not be considered as cache hits.`

Given that pruning is actually implemented, the MAY and SHOULD doesn't make as much sense, but I may be misinterpting. 

Also, let's capitalize NOT as well.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
```
👍

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
+		// alignment between the TTL and the block production rate.
```
```suggestion
		// alignment between the TTL and the block production rate,
		// by accessing onchain block times directly.
```

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	}
+)
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
```
Was there really no OSS library for this?

bryanchriswhite replied:
This work (and #995) is more about interface discovery and laying a solid foundation than implementing an in-memory cache, which isn't the end goal. In fact, we MUST iterate towards a persistent cache as it's necessary in order for off-chain binaries to tolerate restarts; e.g., a relayminer restarting in-between claim and proof windows.

My intuition regarding the persistent cache is that it should be backed by badger (or whatever we're already using for off-chain persistence), making the cache implementation itself, a thin wrapper around the badger API. Additionally, I think we should consider designing an off-chain "multi-store", similar in princple to the on-chain multi-store. The rationale is that we should prefer consolidating off-chain persistence into a single database over maintaining multiple, feature-specific databases. I don't see this as a huge lift given the solid foundational architecture/conventions we've got going (e.g. `context.Context`, `depinject`, etc.).

All things considered, I determined that it didn't make sense to use an external dependency here. Was there a particular library that had caught your attention?

I'd be happy to continue the discussion here or elsewhere, and/or open an issue.

Olshansk replied:
> All things considered, I determined that it didn't make sense to use an external dependency here. Was there a particular library that had caught your attention?

Nope, just wanted to get your thought process!

> I'd be happy to continue the discussion here or elsewhere, and/or open an issue.

I'll request one better.

Why not transfer the explanation you have here into a `pkg/client/query/cache/README.md` with ` TODO_FUTURE` assigned to your name?

bryanchriswhite replied:
> Why not transfer the explanation you have here into a pkg/client/query/cache/README.md with  TODO_FUTURE assigned to your name?

I was compelled to open an issue instead. :sweat_smile: 

- https://github.com/pokt-network/poktroll/issues/1014

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
+// optional historical value support.
+type InMemoryCache[T any] struct {
```
Does this need to be publically exposed

bryanchriswhite replied:
Nope - good catch! :raised_hands:

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	timestamp time.Time
+}
+
+// cacheItemHistory stores cachedItems by height and maintains a sorted list of
```
Awesome.

Beginning to remind me of our interview question a bit 😅

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+// Get retrieves the value from the cache with the given key. If the cache is
+// configured for historical mode, it will return the value at the latest **known**
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height.
```
```suggestion
// guaranteed to be the current height w.r.t the blockchain.
```

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+
+	cItem := item.(cacheItem[T])
+	if c.config.TTL > 0 && time.Since(cItem.timestamp) > c.config.TTL {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
```
So is there ever cleanup? Is it something we could defer or do lazyl elsewhere?

bryanchriswhite replied:
Yes, will #PUC.

bryanchriswhite replied:
```go
	if isTTLEnabled && isCacheItemExpired {
		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
		// otherwise, the read lock would be insufficient. The value will be
		// overwritten by the next call to Set(). If usage is such that values
		// aren't being subsequently set, maxKeys (if configured) will eventually
		// cause the pruning of values with expired TTLs.
		return zero, ErrCacheMiss.Wrapf("key: %s", key)
	}
```

Olshansk replied:
I think the `if isTTLEnabled` part was lost as part of another commit.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+// GetAtHeight retrieves the value from the cache with the given key, at the given
+// height. If a value is not found for that height, the value at the nearest previous
```
> If a value is not found for that height, the value at the nearest previous height is returned.

Thoughts on adding a `nearestOk` bool to make this functionality ultra explicit?

Olshansk replied:
When I see a function like `GetAtHeight`, my immediate thought is that it'll return the value at the height provided (if available) or nil otherwise.

For `height=5`, it's the difference between `select * from table where height=5` and `select * from table where height <= 5`. 

There could be business logic that depends on state transitions AT A SPECIFIC HEIGHT, so I'm just afraid that it's not explicit enough.

bryanchriswhite replied:
I take your point, and agree that his is a valid concern. However, I think we're getting hung up on an implementation detail that is actually an optimization. If the implementation were completely honest, it would be populating the history at **each** height, where the value would be the same as the previous height so long as it wasn't changed (or deleted; we have few of these objects: claims/proofs).

The fact that the heights are omitted here is an optimization which takes advantage of the assumption that this pattern holds due to the blockchain context that we're in. Even in the case of the few onchain objects which do violate this assumption , I don't think it becomes problematic. I also don't think we have a good reason to consider caching claims or proofs yet.

A concrete example would be something like:
1. A claim was created at height 10
2. The claim was queried and cached at height 12
3. The claim was settled and deleted at height 20
4. The claim was queried at height 30, for height 22
5. The claim would be returned even though it was deleted two blocks prior to the queried height

bryanchriswhite replied:
```go
// HistoricalQueryCache extends QueryCache to support getting and setting values
// at multiple heights for a given key.
type HistoricalQueryCache[T any] interface {
	QueryCache[T]
	// GetAsOfHeight retrieves the nearest value <= the specified height
	GetAsOfHeight(key string, height int64) (T, error)
	// SetAsOfHeight adds or updates a value at a specific height
	SetAsOfHeight(key string, value T, height int64) error
}
```

Olshansk replied:
_tl;dr I still disagree but will leave the thread here :)_

There are "implicit assumptions"  in this conversation/thread.

Whenever that happens, I always think it's something that requires context or institutional knowledge. Further, it's something I can see us forgetting as well.

In either case, you are the main implementor and designer here. If this becomes an issue, I'm sure it'll resurface.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	cItem := item.(cacheItem[T])
```
Consider checking if `ok` because this is a type assertion and not just a type cast

https://stackoverflow.com/questions/20494229/what-is-the-difference-between-type-conversion-and-type-assertion

bryanchriswhite replied:
Thanks for the explanation, I'm familiar with the distinction. :blush:

Olshansk replied:
Haha - too many late nights on python bindings ;)

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	cItem := item.(cacheItem[T])
+	if c.config.TTL > 0 && time.Since(cItem.timestamp) > c.config.TTL {
```
```suggestion
	isTTLEnabled := c.config.TTL > 0
	isCacheItemExpired := time.Since(cItem.timestamp) > c.config.TTL
	if isTTLEnabled && isCacheItemExpired {
```

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+		return zero, ErrCacheMiss.Wrapf("key: %s, height: %d", key, getHeight)
+	}
+
+	itemHistory := itemHistoryAny.(cacheItemHistory[T])
```
Ditto regarding assertion & casting.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	item := itemHistory.itemsByHeight[nearestCachedHeight]
+	if c.config.TTL > 0 && time.Since(item.timestamp) > c.config.TTL {
```
Consider using the same language here as above

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+		return zero, ErrCacheMiss.Wrapf("key: %s, height: %d", key, getHeight)
+	}
+
+	item := itemHistory.itemsByHeight[nearestCachedHeight]
```
Should we check if the item exists?

Olshansk replied:
Consider consolidating `item`/`value` language.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+		return c.SetAtHeight(key, value, c.latestHeight.Load())
+	}
+
+	if c.config.MaxKeys > 0 && int64(len(c.items)) >= c.config.MaxKeys {
```
```suggestion
	maxKeysEnabled := c.config.MaxKeys > 0
	maxKeysReached := int64(len(c.items)) >= c.config.MaxKeys
	if maxKeysEnabled &&  maxKeysReached {
```

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	// | --------------- | ------------------------------ |
+	// | false           | map[string]cacheItem[T]        |
+	// | true            | map[string]cacheItemHistory[T] |
+	items map[string]any
```
Why items instead of values?

bryanchriswhite replied:
Will #PUC.

bryanchriswhite replied:
I restructured things a bit and did some renaming:

```go
// inMemoryCache provides a concurrency-safe in-memory cache implementation with
// optional historical value support.
type inMemoryCache[T any] struct {
	config       queryCacheConfig
	latestHeight atomic.Int64

	// valuesMu is used to protect values AND valueHistories from concurrent access.
	valuesMu sync.RWMutex
	// values holds the cached values in non-historical mode.
	values map[string]cacheValue[T]
	// valueHistories holds the cached historical values in historical mode.
	valueHistories map[string]cacheValueHistory[T]
}

// cacheValue wraps cached values with a timestamp for later comparison against
// the configured TTL.
type cacheValue[T any] struct {
	value     T
	timestamp time.Time
}

// cacheValueHistory stores cachedItems by height and maintains a sorted list of
// heights for which cached items exist. This list is sorted in descending order
// to improve performance characteristics by positively correlating index with age.
type cacheValueHistory[T any] struct {
	// sortedDescHeights is a list of the heights for which values are cached.
	// It is sorted in descending order.
	sortedDescHeights []int64
	heightMap         map[int64]cacheValue[T]
}
```

Olshansk replied:
This is great, ty!

Olshansk commented on pkg/client/query/cache/memory_test.go:
```diff
@@ -0,0 +1,339 @@
+package cache
```
@bryanchriswhite I'm not reviewing this but lmk if there's a particular section in the tests you'd want me to take a look at.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,304 @@
+	c.itemsMu.Lock()
+	defer c.itemsMu.Unlock()
+
+	delete(c.items, key)
```
What if we delete the (key,value) pair for `latestHeight`?

bryanchriswhite replied:
Could you elaborate, I'm not quite following.

Olshansk replied:
Yea, this comment doesn't make sense if you don't know where I was looking or what I was thinking...

If we delete the key for the latest height, do we need to also make an appropriate call to `c.latestHeight.Store(??)`?

If this is indeed the case, please add a unit test for it.

If not, would just appreciate a quick (no need for anything big) explanation.

bryanchriswhite replied:
Gotcha, good question. I don't think we need to update `c.latestHeight`, regardless of which key is being deleted.  For `Get`s, this should be completely transparent to the caller; for `Set`s, it means that if you call `Delete`, followed by `Set` (as opposed to `SetAtHeight`), then the height you're setting will be the last `c.latestHeight`. I think this behavior is correct. 

> If not, would just appreciate a quick (no need for anything big) explanation.
I did see :point_up_2: but had to think it through for myself anyways:

```mermaid
flowchart LR

subgraph hc2[Historical Cache]
    lh2["latestHeight: 2"]
    subgraph k2_2h["key2 history"]
        k2_2h1["height: 1; value 100"]
    end
    subgraph k2_1h["key1 history"]
        k2_1h1["height: 1; value 10"]
        k2_1h2["height: 2; value 20"]
    end
end

subgraph hc3[Historical Cache]
    lh3["latestHeight: 2"]
    subgraph k3_2h["key2 history"]
        k3_2h1["height: 1; value 100"]
    end
end


subgraph hc4[Historical Cache]
    lh4["latestHeight: 2"]
    subgraph k4_1h["key1 history"]
        k4_1h2["height: 2; value 30"]
    end
end

hc2 --"Delete(key1)"--> hc3
hc3 --"Set(key1, 30)"--> hc4

lh2 --> k2_1h2
lh4 --> k4_1h2
```

Olshansk replied:
Given the explanation and diagram, I don't fully understand the need for `latestHeight` at all. In fact, I think we should delete it now if the behaviour you described is expected.

bryanchriswhite replied:
I neglected to add a comment to this thread but after discussing with @red-0ne, we decided that it should be an error to call `Set()` on a `HistoricalQueryCache`. I will update the interface and `inMemoryCache#Set()` to reflect this. In this context, `latestVersion` is no longer necessary and can be removed.

bryanchriswhite replied:
```
// QueryCache is a key/value store style interface for a cache of a single type.
// It is intended to be used to cache query responses (or derivatives thereof),
// where each key uniquely indexes the most recent query response.
type QueryCache[T any] interface {
	Get(key string) (T, error)
	Set(key string, value T) error
	Delete(key string)
	Clear()
}

// HistoricalQueryCache extends QueryCache to support getting and setting values
// at multiple heights for a given key.
type HistoricalQueryCache[T any] interface {
	// GetLatestVersion retrieves the historical value with the highest version number.
	GetLatestVersion(key string) (T, error)
	// GetAsOfVersion retrieves the nearest value <= the specified version number.
	GetAsOfVersion(key string, version int64) (T, error)
	// SetAsOfVersion adds or updates a value at a specific version number.
	SetAsOfVersion(key string, value T, version int64) error
}
```
(class diagrams also updated)

Olshansk commented on pkg/client/query/cache/memory_test.go:
```diff
@@ -0,0 +1,339 @@
+		require.ErrorIs(t, err, ErrCacheMiss)
+	})
+
+	t.Run("max size eviction", func(t *testing.T) {
```
```suggestion
	t.Run("max key count eviction", func(t *testing.T) {
```

Olshansk reviewed (changes requested):
@bryanchriswhite I did a superficial review but did not dive into the validation of the business logic line-by-line.

Is there any section where you'd want another pair of 👀 ?

bryanchriswhite commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,69 @@
+type queryCacheConfig struct {
+	// MaxKeys is the maximum number of items (key/value pairs) the cache can
+	// hold before it starts evicting.
+	MaxKeys        int64
```
Good catch - thanks!

bryanchriswhite commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,69 @@
+	MaxKeys        int64
+	EvictionPolicy EvictionPolicy
+	// TTL is how long items should remain in the cache. Items older than the TTL
+	// MAY not be evicted but SHOULD not be considered as cache hits.
```
> What determines if it will or will not be evicted? MAY NOT is a bit confusing.

The implementation determines that, this config doesn't know anything about the implementation.
In the case of `InMemoryCache`, eviction only happens when on a per-key/value basis, when a key/value is updated.

> Should we have a WILL NOT be considered cache hits? I can't tell what the behaviour is with SHOULD NOT

:+1: Agreed.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -10,116 +10,119 @@ import (
 	// It is sorted in descending order.
 	sortedDescHeights []int64
-	itemsByHeight     map[int64]cacheItem[T]
+	heightMap         map[int64]cacheValue[T]
```
I was considering `heightCacheValueMap`, but `heightMap` just doesn't feel like enough.
```suggestion
	heightValueMap         map[int64]cacheValue[T]
```

bryanchriswhite replied:
Continuing the direction of https://github.com/pokt-network/poktroll/pull/994#discussion_r1890025275, I'm going with `heightVersionsMap` here.

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -10,116 +10,119 @@ import (
 
-// cacheItemHistory stores cachedItems by height and maintains a sorted list of
+// cacheValueHistory stores cachedItems by height and maintains a sorted list of
 // heights for which cached items exist. This list is sorted in descending order
```
```suggestion
// heights for which cached values exist. This list is sorted in descending order
```

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -134,37 +137,47 @@ func (c *InMemoryCache[T]) GetAtHeight(key string, getHeight int64) (T, error) {
+		return zero, ErrCacheInternal.Wrapf("failed to load historical value for key: %s, height: %d", key, getHeight)
+	}
+
+	if c.config.ttl > 0 && time.Since(value.cachedAt) > c.config.ttl {
```
Looks like `isTTLEnabled` dissapeared: https://github.com/pokt-network/poktroll/pull/994#discussion_r1886643319

red-0ne commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+	}
+
+	// Prune historical values for this key, where the setHeight
+	// is oder than the configured numHistoricalValues.
```
```suggestion
	// is older than the configured numHistoricalValues.
```

bryanchriswhite replied:
:hankey:  :joy:

red-0ne commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+			// (oldest) to highest (youngest) height, once we encounter a cachedHeight
+			// that is younger than the configured numHistoricalValues, ALL subsequent
+			// heights SHOULD also be younger than the configured numHistoricalValues.
+			if setHeight-cachedHeight <= c.config.numHistoricalValues {
```
I struggle to make sense of this.

As of my understanding `config.numHistoricalValues` would be the "size" of the historical values retained. If so, the condition to delete an entry would be:
```
if
  len(valueHistory.sortedDescHeights) > config.numHistoricalValues &&
  setHeight.isYoungerThan(oldestCachedHeight)
```

bryanchriswhite replied:
Great catch - thank you! :raised_hands:

red-0ne reviewed (commented):
Great caching design.

Left some comments but did not tend to `pkg/client/query/cache/memory_test.go` yet. Which I'll do right after this one.

bryanchriswhite commented on pkg/client/query/cache/config.go:
```diff
@@ -13,57 +13,77 @@ const (
+// WithHistoricalMode enables historical caching with the given numHistoricalValues
 // configuration; if 0, no historical pruning is performed.
-func WithHistoricalMode(pruneOlderThan int64) QueryCacheOptionFn {
+func WithHistoricalMode(numHistoricalBlocks int64) QueryCacheOptionFn {
```
I intentionally chose `numHistoricalBlocks` over `numHistoricalValues` to disambiguate "key/value" from "historical value" and to imply the frequency. I suppose it's not quite appropriate to imply the frequency at this level, as it's out of the scope this component's concerns. After a chat w/ @red-0ne I would rename it to `numRetainedVersions`. Wdyt?

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -10,116 +10,119 @@ import (
 }
 
-// cacheItemHistory stores cachedItems by height and maintains a sorted list of
+// cacheValueHistory stores cachedItems by height and maintains a sorted list of
```
That was the last one. :sweat_smile:

bryanchriswhite commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,89 @@
+		return ErrQueryCacheConfigValidation.Wrapf("eviction policy %d not imlemented", cfg.evictionPolicy)
+	}
+
+	return nil
```
```go
	if cfg.numRetainedVersions > 0 && cfg.historical == false {
		return ErrQueryCacheConfigValidation.Wrap("numRetainedVersions > 0 requires historical mode to be enabled")
	}

	if cfg.historical && cfg.numRetainedVersions < 0 {
		return ErrQueryCacheConfigValidation.Wrapf("numRetainedVersions MUST be >= 0, got: %d", cfg.numRetainedVersions)
	}
```

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+	c.valuesMu.RLock()
+	defer c.valuesMu.RUnlock()
+
+	valueHistory := c.valueHistories[key]
```
:pray:

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+	}
+
+	isMaxKeysConfigured := c.config.maxKeys > 0
+	cacheHasMaxKeys := int64(len(c.values)) >= c.config.maxKeys
```
Great catch - thank you! :raised_hands:

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+// guaranteed to be the current height.
+func (c *inMemoryCache[T]) Set(key string, value T) error {
+	if c.config.historical {
+		return c.SetAtHeight(key, value, c.latestHeight.Load())
```
I think this is more about the specific behavior of `Set()` in historical mode. This is indeed an edge case, and I don't think we should find ourselves doing this unless for a very specific reason. In such a case, I think it's more accurate to say that instead of `fetchValue()` updating the latest version (height), it would be an intermediate call to `SetAsOfVersion()` where the version is newer than the current latest version. I think this is the scenario I outlined above: https://github.com/pokt-network/poktroll/pull/994#discussion_r1889842645.

bryanchriswhite replied:
As discussed, I've disabled `#Set()` in historical mode.

red-0ne commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,378 @@
+	if version > latestVersion {
+		// NB: Only update if c.latestVersion hasn't changed since we loaded it above.
+		if c.latestVersion.CompareAndSwap(latestVersion, version) {
+			latestVersion = version
```
```suggestion
			latestVersion = version
			c.latestVersion.Store(version)
```

bryanchriswhite replied:
Can you elaborate on the thinking here? In this case, the `latestVersion` was updated to an even "newer/later" concurrently. The point of using `#CompareAndSwap()` here was to avoid the latest version from regressing.

red-0ne replied:
My bad. I overlooked the fact that `CompareAndSwap` do swap too :sweat_smile:

red-0ne commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,362 @@
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height w.r.t the blockchain.
+func (c *inMemoryCache[T]) Get(key string) (T, error) {
+	if c.config.historical {
```
Looking at this function it was not clear why this conditional block is excluded from the `valuesMu` lock.

Now I know that it `GetAsOfVersion` is locking too, I could no longer justify placing #PUCing it :sweat_smile:

red-0ne reviewed (commented):
Asking for a small change, but otherwise LGTM.

Not approving yet as I want review the tests too.

Olshansk commented on pkg/client/interface.go:
```diff
@@ -375,8 +375,8 @@ type QueryCache[T any] interface {
-	// SetAtHeight adds or updates a value at a specific height
-	SetAtHeight(key string, value T, height int64) error
+	// GetAsOfVersion retrieves the nearest value <= the specified version number.
+	GetAsOfVersion(key string, version int64) (T, error)
```
Can we go with one of the following instead:

1. GetVersioned
2. GetAtVersion
3. GetHistorical

Don't care which one but `AsOf` in a function name just doesn't feel right.

bryanchriswhite replied:
In that case, `GetVersion` would be my preference. Any objections @red-0ne @Olshansk?

bryanchriswhite replied:
Also adding `GetLatestVersion`.

Olshansk replied:
@bryanchriswhite Can we add `GetLtestVersino`

Olshansk commented on pkg/client/query/cache/config.go:
```diff
@@ -13,57 +13,86 @@ const (
+
+	// evictionPolicy determines which values are removed when number of keys in the cache reaches maxKeys.
+	evictionPolicy EvictionPolicy
+	// ttl is how long values should remain valid in the cache. Items older than the
```
This updated comment is great

Olshansk commented on pkg/client/query/cache/config.go:
```diff
@@ -13,57 +13,86 @@ const (
-	// values. If 0, no historical pruning is performed. It only applies when
-	// historical is true.
-	pruneOlderThan int64
+	// maxVersionAge is the max difference between the latest known version and
```
this sentence is hard to grok.

Any chance you could improve it with an example?

Olshansk commented on pkg/client/query/cache/memory.go:
```diff
@@ -10,295 +10,369 @@ import (
-	items map[string]any
+type inMemoryCache[T any] struct {
+	config        queryCacheConfig
+	latestVersion atomic.Int64
```
Per my other comment, can you #PUC why we need this and how its used?

bryanchriswhite replied:
:saluting_face: :sweat_smile:

Olshansk commented:
@bryanchriswhite Could we get this over the finish line ASAP so it can be used in #1050?

red-0ne reviewed (commented):
Left few comments but it feels like the current implementation is calling for having an interface and implementation for each type of cache (historical, non-historical).

It will remove a lot of conditional branching and runtime errors.

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,412 @@
+
+	cachedValue, exists := c.values[key]
+	if !exists {
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
```
We can't literally return `cachedValue` because it is not of type `T`, it is a `cacheValue[T]`, the `#value` field of which would have our `T` type; however, since `cachedValue` is `nil`, we cannot access it.

bryanchriswhite commented on pkg/client/query/cache/memory.go:
```diff
@@ -0,0 +1,412 @@
+		var oldestKey string
+		var oldestTime time.Time
+		for key, valueHistory := range c.valueHistories {
+			mostRecentVersion := valueHistory.sortedDescVersions[0]
```
I think we only disagree about the word "recent". I was seeing it as interchangeable with "latest", where both would refer to the "highest version number".

bryanchriswhite replied:
I think we only disagree about the word "recent". I was seeing it as interchangeable with "latest", where both would refer to the "highest version number".

red-0ne commented on pkg/cache/memory/historical_kvcache_test.go:
```diff
@@ -0,0 +1,179 @@
+		require.False(t, isCached)
+	})
+
+	t.Run("historical cache ignores TTL expiration", func(t *testing.T) {
```
Historical cache is still subject to `maxKeys` FIFO eviction policy.

I believe it's worth test covering it.

Olshansk reviewed (changes requested):
Just realized I never published my review from yesterday :(

bryanchriswhite commented on pkg/cache/interface.go:
```diff
@@ -0,0 +1,22 @@
+	Clear()
+}
+
+// HistoricalKeyValueCache extends KeyValueCache to support getting and setting values
```
```go
// HistoricalKeyValueCache is a key/value store style interface for a cache of a single type.
// It is intended to be used to cache arbitrary data, where each key uniquely indexes
// a mapping of version numbers to values corresponding to the historical values of the data
// associated that key.
```

bryanchriswhite commented on pkg/cache/errors.go:
```diff
@@ -0,0 +1,10 @@
+
+import "cosmossdk.io/errors"
+
+const codesace = "client/query/cache"
```
Nice catch! :raised_hands: 

Given that it's no longer in the `client` pkg, I'm going with:
```go
const codespace = "cache"
```

bryanchriswhite commented on pkg/cache/memory/kvcache.go:
```diff
@@ -0,0 +1,150 @@
+		delete(c.values, oldestKey)
+		return
+
+	case LeastRecentlyUsed:
```
I moved the comment above this line and rephrased slightly:
```go
	// DEV_NOTE: The following cases SHOULD NEVER happen, KeyValueCacheConfig#Validate, SHOULD prevent it.
	case LeastRecentlyUsed:
```

bryanchriswhite commented on pkg/cache/memory/historical_kvcache.go:
```diff
@@ -0,0 +1,233 @@
+
+// GetVersion retrieves the value from the cache with the given key, as of the
+// given version. If a value is not found for that version, the value at the nearest
+// previous version is returned. If the cache is not configured for historical mode,
+// it returns an error.
+func (c *historicalKeyValueCache[T]) GetVersion(key string, version int64) (T, bool) {
```
Thank you! :heart:

bryanchriswhite commented on pkg/cache/memory/historical_kvcache.go:
```diff
@@ -0,0 +1,233 @@
+// NewHistoricalKeyValueCache creates a new historicalKeyValueCache with the
+// configuration generated by the given option functions.
+func NewHistoricalKeyValueCache[T any](opts ...KeyValueCacheOptionFn) (*historicalKeyValueCache[T], error) {
+	config := DefaultHistoricialKeyValueCacheConfig
```
:pray:

bryanchriswhite commented on pkg/cache/memory/historical_kvcache.go:
```diff
@@ -0,0 +1,233 @@
+		return zero, false
+	}
+
+	return value.value, true
```
I went with `cachedValue` for consistency.

bryanchriswhite commented on pkg/cache/memory/config.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// Validate ensures that the historicalKeyValueCacheConfig isn't configured with incompatible options.
+func (cfg *historicalKeyValueCacheConfig) Validate() error {
```
```go
// Validate ensures that the historicalKeyValueCacheConfig isn't configured with incompatible options.
func (cfg *historicalKeyValueCacheConfig) Validate() error {
	if err := cfg.keyValueCacheConfig.Validate(); err != nil {
		return err
	}
...
```

bryanchriswhite commented on pkg/cache/memory/historical_kvcache.go:
```diff
@@ -0,0 +1,233 @@
+// returns an error.
+func (c *historicalKeyValueCache[T]) SetVersion(key string, value T, version int64) error {
+	// DEV_NOTE: MUST call getLatestVersionNumber() before locking valuesMu.
+	latestVersion := c.getLatestVersionNumber(key)
```
Agreed.

bryanchriswhite commented on pkg/client/query/cache/config.go:
```diff
@@ -0,0 +1,100 @@
+	historical bool
+	// maxVersionAge is the max difference between the latest known version and
+	// any other version, below which value versions are retained, and above which
+	// value versions are pruned.
```
All of the above ("pruned" == "evicted" == "deleted"). As it states, they ARE pruned. The word "evict" has only been used so far in a slightly different context, related to the `MaxKeys` config field.

Do you have a specific suggestion on how to improve clarity here?

Olshansk replied:
```suggestion
// value versions are pruned (i.e. evicted).
```

Olshansk reviewed (commented):
@bryanchriswhite I did a "show changes since your last review" and it seems like all the files have changed?

Either way, lmk when this is ready for a re-review.

bryanchriswhite commented:
Based on prior feedback and the continued iteration here, I've refactored `GetVersion()` to ONLY return the specific version requested, and added a new `GetVersionLTE()` method which implements the previous behavior:
```
	// GetVersionLTE retrieves the value at the nearest version <= maxVersion.
	// It returns true if a value meeting the version criteria for the given
	// key is cached, false otherwise.
	GetVersionLTE(key string, maxVersion int64) (T, bool)
```

red-0ne reviewed (approved):
Love the shape it took :fire:

Thanks for your patience iterating over it.

Olshansk reviewed (approved):
Love the final shape of this!

bryanchriswhite commented:
PLEASE DO NOT DELETE THIS BRANCH UNTIL DOWNSTREAM PRS HAVE BEEN REBASED
